core/commands/pin.go

```diff
@@ -195,60 +208,43 @@ Example:
 			return
 		}
 
-		typeStr, found, err := req.Option("type").String()
+		typeStr, typeStrFound, err := req.Option("type").String()
 		if err != nil {
 			res.SetError(err, cmds.ErrNormal)
 			return
 		}
-		if !found {
-			typeStr = "recursive"
-		}
 
-		switch typeStr {
-		case "all", "direct", "indirect", "recursive":
-		default:
-			err = fmt.Errorf("Invalid type '%s', must be one of {direct, indirect, recursive, all}", typeStr)
-			res.SetError(err, cmds.ErrClient)
-		}
-
-		keys := make(map[string]RefKeyObject)
-		if typeStr == "direct" || typeStr == "all" {
-			for _, k := range n.Pinning.DirectKeys() {
-				keys[k.B58String()] = RefKeyObject{
-					Type: "direct",
-				}
+		if typeStrFound {
+			switch typeStr {
+			case "all", "direct", "indirect", "recursive":
+			default:
+				err = fmt.Errorf("Invalid type '%s', must be one of {direct, indirect, recursive, all}", typeStr)
+				res.SetError(err, cmds.ErrClient)
+				return
 			}
 		}
-		if typeStr == "indirect" || typeStr == "all" {
-			ks := key.NewKeySet()
-			for _, k := range n.Pinning.RecursiveKeys() {
-				nd, err := n.DAG.Get(n.Context(), k)
-				if err != nil {
-					res.SetError(err, cmds.ErrNormal)
-					return
-				}
-				err = dag.EnumerateChildren(n.Context(), n.DAG, nd, ks)
-				if err != nil {
-					res.SetError(err, cmds.ErrNormal)
-					return
-				}
 
+		var keys map[string]RefKeyObject
+
+		if len(req.Arguments()) > 0 {
+			if !typeStrFound {
+				typeStr = "all"
 			}
-			for _, k := range ks.Keys() {
-				keys[k.B58String()] = RefKeyObject{
-					Type: "indirect",
-				}
-			}
-		}
-		if typeStr == "recursive" || typeStr == "all" {
-			for _, k := range n.Pinning.RecursiveKeys() {
-				keys[k.B58String()] = RefKeyObject{
-					Type: "recursive",
-				}
+
+			keys, err = pinLsKeys(req.Arguments(), typeStr, req.Context(), n)
+		} else {
+			if !typeStrFound {
+				typeStr = "recursive"
```

Comment:
Yeah, it would be more logical for the default to be the same in both cases, but when there are args it is natural to have the default be "all" and setting it to "all" when there are no args is a bit backward incompatible, so I preferred to avoid it.

Now if you are ok for the default being "all" in both case I can change that.